kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaEvictionService.java

```diff
@@ -0,0 +1,164 @@
+
+  @VisibleForTesting protected int futuresListTimeoutSecs = DEFAULT_ZK_TIMEOUT_SECS;
+
+  public static final String REPLICA_MARK_EVICT_SUCCEEDED = "replica_mark_evict_succeeded";
+  public static final String REPLICA_MARK_EVICT_FAILED = "replica_mark_evict_failed";
+  public static final String REPLICA_MARK_EVICT_TIMER = "replica_mark_evict_timer";
+
+  protected final Counter replicaMarkEvictSucceeded;
+  protected final Counter replicaMarkEvictFailed;
+  private final Timer replicaMarkEvictTimer;
+
+  public ReplicaEvictionService(
+      CacheSlotMetadataStore cacheSlotMetadataStore,
+      ReplicaMetadataStore replicaMetadataStore,
+      KaldbConfigs.ManagerConfig managerConfig,
+      MeterRegistry meterRegistry) {
+    this.cacheSlotMetadataStore = cacheSlotMetadataStore;
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.managerConfig = managerConfig;
+    this.meterRegistry = meterRegistry;
+
+    // schedule configs checked as part of the AbstractScheduledService
+
+    replicaMarkEvictSucceeded = meterRegistry.counter(REPLICA_MARK_EVICT_SUCCEEDED);
+    replicaMarkEvictFailed = meterRegistry.counter(REPLICA_MARK_EVICT_FAILED);
+    replicaMarkEvictTimer = meterRegistry.timer(REPLICA_MARK_EVICT_TIMER);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    markReplicasForEviction();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getReplicaEvictionServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting replica eviction service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closed replica eviction service");
+  }
+
+  /**
+   * Marks cache slots that contain an expired replica as ready for eviction. Once eviction has
+   * started the cache node will further update the state indicating it has started evicting the
+   * replica from memory/disk.
+   */
+  protected int markReplicasForEviction() {
+    Timer.Sample evictionTimer = Timer.start(meterRegistry);
+
+    Map<String, ReplicaMetadata> replicaMetadataByReplicaId =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            .collect(Collectors.toUnmodifiableMap(ReplicaMetadata::getName, Function.identity()));
+
+    long expireAfterNow = Instant.now().toEpochMilli();
+    AtomicInteger successCounter = new AtomicInteger(0);
+    List<ListenableFuture<?>> replicaEvictions =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            // get all the slots that are currently assigned, but have an expiration in the past
+            .filter(
+                cacheSlotMetadata ->
+                    cacheSlotMetadata.cacheSlotState.equals(
+                            Metadata.CacheSlotMetadata.CacheSlotState.ASSIGNED)
```

Comment:
We should be evicting all snapshots which are in LIVE, not ASSIGNED. Since the node stays assigned for a short amount of time, it's probably ok to let it go to loading. Should we also evict snapshots in LOADING state?

Also, extract the logic for the filter into a predicate so the we can test this logic easily.

Reply:
Ah, good call out - these should all be checking for LIVE, not ASSIGNED. Also extracted the logic for the filter.

I think leaving the ones marked LOADING is fine - the only time we would expect a node to be LOADING an expired snapshot is if it had somehow gotten "stuck" for a prolonged period of time - which the monitoring service should address.